compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -914,6 +914,22 @@ impl<'hir> LoweringContext<'_, 'hir> {
         hir::ExprKind::Block(&self.block_all(whole_span, stmts, None), None)
     }
 
+    /// Returns whether a given expression resolves to a tuple struct.
+    /// This is used in lowering destructuring assignments for better diagnostics.
+    fn is_tuple_struct(&mut self, expr: &Expr) -> bool {
+        // For tuple struct destructuring, it must be a non-qualified path (like in patterns).
+        if !matches!(expr.kind, ExprKind::Path(None, _)) {
+            return false;
+        }
+        // Does the path resolves to something allowed in a tuple struct/variant pattern?
+        let partial_res = self.resolver.get_partial_res(expr.id);
+        if let Some(partial_res) = partial_res {
+            return partial_res.unresolved_segments() == 0
+                && partial_res.base_res().expected_in_tuple_struct_pat();
+        }
+        false
```

Comment:
The logic should be
```rust
        if let Some(partial_res) = partial_res {
            if partial_res.unresolved_segments() == 0 && !partial_res.base_res().expected_in_tuple_struct_pat() {
                return false;
            }
        }
        true
```
, i.e. if we know for sure that it's not a tuple struct, then we treat it as a regular assignment.
If we are *not sure* (e.g. `partial_res.unresolved_segments() > 0`), then it may be resolved as a tuple struct during type checking, and we need to treat it as destructuring assignment.
This is a "fast reject" mechanism, not an "accept" mechanish.
I wrote about this in https://github.com/rust-lang/rust/pull/78836#discussion_r519414768.

Reply:
Thanks for explaining it again. I must have misunderstood your previous comment. I've changed to now.